pdns/dnsname.cc

```diff
@@ -808,6 +808,23 @@ bool DNSName::RawLabelsVisitor::empty() const
   return d_position == 0;
 }
 
+bool DNSName::matches(const std::string_view& wire_uncompressed) const
+{
+  if (wire_uncompressed.empty() != empty() || wire_uncompressed.size() < d_storage.size()) {
+    return false;
+  }
+
+  const auto* our = d_storage.cbegin();
+  const auto* other = wire_uncompressed.cbegin();
+  for (; our != d_storage.cend() && other != wire_uncompressed.cend(); ++our, ++other) {
+    if (dns_tolower(*other) != dns_tolower(*our)) {
+      return false;
+    }
+  }
+
+  return our == d_storage.cend();
```

Comment:
This can probably be simplified into something like this (untested):
```suggestion
  // equivalent to pdns_iequals(wire_uncompressed, d_storage.substr(0, wire_uncompressed.size());
  return pdns_ilexicographical_compare_three_way(wire_uncompressed, std::string_view(d_storage.data(), wire_uncompressed.size())) == 0;
```

Reply:
I'll look into it but I'm not sure this will work, because `wire_uncompressed` may contain more than just the wire representation of the qname. In the dnsdist case it contains the whole response packet minus the `dnsheader`. I should document that clearly, at the very least.

Reply:
I added a comment explaining what `DNSName::matches` accepts, and confirmed that I cannot use `pdns_ilexicographical_compare_three_way` without important changes.

Reply:
I am a bit surprised by this, because what you are doing is a case insensitive comparison on the smallest of the two lengths. But that can be refined later.

Reply:
Oh, I think I did not understand that you were suggesting creating a new `string_view`, which makes a lot of sense. Let me take a new look at it.